pkg/compactor/bucket_compactor.go

```diff
@@ -806,6 +812,16 @@ func (c *BucketCompactor) Compact(ctx context.Context, maxCompactionTime time.Du
 			return err
 		}
 
+		// Record the difference between now and the max time for a block being compacted. This
+		// is used to detect compactors not being able to keep up with the rate of blocks being
+		// created. The idea is that most blocks should be for within 24h or 48h.
+		now := time.Now().Unix()
+		for _, gr := range jobs {
+			for _, meta := range gr.Metas() {
+				c.metrics.blocksMaxTimeDelta.Observe(float64(now - meta.MaxTime))
```

Comment:
How do you plan to use this new metric? Could you show a sample query to alert on?

I'm asking because the `BucketCompactor.Compact()` is called for each tenant. If a tenant is lagging behind, but other tenants are good, you will track a timestamp for an old block every while and then, but in the alert query may be difficult to understand if that issue is resolved or not (it may not be resolved, but not even tracked until that tenant blocks are compacted again).

I'm not saying this approach is wrong, but I think reviewing this without having the context of how this metric will be queried is a bit difficult.

Reply:
The query used would be pretty simple, making sure that the p75 of deltas were less than 48h (not sure what the actually percentile or delta value would be, I need to test this first). For the alert, maybe something like:

```
histogram_quantile(0.75, sum by (le) (rate(cortex_compactor_block_max_time_delta_seconds_bucket{namespace="$namespace"}[$__rate_interval]))) / 3600 > 48
```

You are correct, this wouldn't help identify when an individual tenant is lagging but most others are not. My thinking was that this would be useful for detecting when compactors are under-scaled for a cell. For example, I suspect the recent performance issues in prod-10 would have been caught by something like this.

Reply:
Right. We've never been able to come up with a very good metric (yet). I think this new metric can be useful in some scenarios, maybe not much in others, but until we'll have a better way to actually measure how much the compactor is lagging behind for real, I'm 👍 to add it.